rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+  - e.g. `.add_system(player_controls.before(GameLabels::Physics).after(CoreLabels::Input))`
+- it may have one or more **run criteria** attached
+  - a system is only executed if all of its run criteria return `true`
+  - **states** are a special, more complex pattern that use run criteria to determine whether or not a system should run in a current state
+  - e.g. `.add_system(physics.run_if_resource_equals(EnablePhysics(true)))`
+- it may have one or more **labels**, allowing other systems to refer to it and enabling mass-configuration
+  - e.g. `.add_system(physics.label(GameLabels::Physics)`
+
+This configuration is additive: adding another ordering constraint, run criteria or label does not replace the existing configuration.
+
+System configuration can be stored in a `SystemConfig` struct. This can be useful to reuse, compose and quickly apply complex configuration strategies, before applying these strategies to labels or individual systems.
+
+### Label configuration
+
+Each system label has its own associated `SystemConfig`, stored in the corresponding `Schedule`.
+When a label is configured, that configuration is passed down in an additive fashion to each system with that label.
+
+You can apply the same label(s) to many systems at once using the `App::add_systems(systems: impl SystemIterator)` method.
+
+```rust
+#[derive(SystemLabel)]
+enum Physics {
+    Forces,
+    CollisionDetection,
+    CollisionHandling,
+}
+
+impl Plugin for PhysicsPlugin{
+
+    fn build(app: &mut App){
+        // Within each frame, physics logic needs to occur after input handling, but before rendering
+        let mut common_physics_config = SystemConfig::new().after(CoreLabel::InputHandling).before(CoreLabel::Rendering);
+
+        app
+        // We can reuse this shared configuration on each of our labels
+        .configure_label(Physics::Forces.configure(common_physics_config).before(Physics::CollisionDetection))
+        .configure_label(Physics::CollisionDetection.configure(common_physics_config).before(Physics::CollisionHandling))
+        .configure_label(Physics::CollisionHandling.configure(common_physics_config))
+        // And then apply that config to each of the systems that have this label
+        .add_system(gravity.label(Physics::Forces))
+        // These systems have a linear chain of ordering dependencies between them
+        // Systems earlier in the chain must run before those later in the chain
+        // Other systems can run in between these systems
+        .add_systems((broad_pass, narrow_pass).chain().label(Physics::CollisionDetection))
+        // Add multiple systems as once to reduce boilerplate!
+        .add_systems((compute_forces, collision_damage).label(Physics::CollisionHandling));
+    }
+}
+```
+
+Labels cannot themselves be labelled: composition is generally clearer than ad-hoc inheritance, and this allows us to compute the configuration for each system more quickly.

```

Comment:
I was wrong on Discord about hierarchical labeling limiting you to one label per node. It would be okay for nodes to have multiple labels if we allowed labeling labels.

Existing system graphs can always be seen as hierarchies because labels *are* subgraphs and the following properties must hold:
- Labels can't be split by nodes that don't have them. / Subgraphs can't be split by nodes that aren't in them.
- Nodes can have more than one label, but labels that share nodes cannot have a constraint between them (or conflicting constraints with other nodes). / If subgraphs share nodes, those subgraphs must be concurrent.

People do intuitively visualize and describe app logic as hierarchies. i.e.
- "movement comes before physics"
- "movement and physics are both under simulation"
- "under physics, we have collision detection followed by constraint solving and dynamics"

We should allow this. If users can order *and* label labels, they can describe these relationships directly, instead of picking random systems or labels to imply them.

It also shows that it's more sound to attach run criteria to labels, run them *once*, and toggle the entire group of systems. A label is a node embedded in a higher-level graph and we can check if it should run. For the actual implementation, if a ready system is the first node we're seeing (with certain labels / in certain subgraphs), we:
1. Execute the run criteria of its never-before-seen labels in hierarchical order.
2. Stop if the RCs of a certain level in the hierarchy don't all return true.
3. Mark the current system, along with all other systems that had non-passing labels as complete (skipped). It must be true that none of them were running or had already run.

Overall, should just involve another bitset to check.

Please elaborate on your "composition vs. inheritance" argument with an example that demonstrates a clear footgun.

This should have no impact on if-needed constraints impl. We can still flatten everything to system-system edges (since we're flattening anyway, I wouldn't make speed claims) and hand the executor a topsorted list of systems. However, it should make (the now future) impl of atomic groups easier.

Reply:
Okay, this is making sense to me. After writing out the "no labelling labels" patterns, I became pretty uncomfortable with the workaround pattern, especially with respect to nesting third-party plugins inside your own structure.

IMO we should re-add labelling labels. And then have a nice rationale section on this whole debate.

Reply:
The run criteria stuff is very interesting. I'll try out a properly rigorous write-up and see how it looks.

Reply:
I guess a relevant issue is that while labels/subgraphs obey those properties I mentioned, systems aren't automatically recognized as being in them, which pseudo-breaks the "subgraphs can't be split by non-members" axiom.

In both cases in the example below, system `C` effectively joins the "physics-collisions-broadphase" subgraph since it's sandwiched between systems `A` and `B`, but only in effect. I don't think someone would ever rely on `C` *not* being considered part of the "p-c-b" subgraph, but we aren't filling out its missing info.

without hierarchical labeling 
```rust
App::new()
    .add_system(
        systemA
            // could put these labels on a reusable SystemConfig, doesn't have to be this verbose
            .with_label("physics") 
            .with_label("collisions")
            .with_label("broadphase")
            .with_label("systemA")
    )
    .add_system(
        systemB
            .with_label("physics")
            .with_label("collisions")
            .with_label("broadphase")
            .with_label("systemB")
    )
    .add_system(
        systemC
            .with_label("systemC")
            .after("systemA")
            .before("systemB")
    )
    .run()
```

with hierarchical labeling
```rust
App::new()
    .add_label("physics")
    .add_label("collisions".with_label("physics"))
    .add_label("broadphase".with_label("collisions"))
    .add_system(
        systemA
            .with_label("broadphase")
            .with_label("systemA")
    )
    .add_system(
        systemB
            .with_label("broadphase")
            .with_label("systemB")
    )
    .add_system(
        systemC
            .with_label("systemC")
            .after("systemA")
            .before("systemB")
    )
    .run()
```

Reply:
> Maybe that's just another thing reported by the ambiguity checker.

Something like, "prior to flattening, report A and B if they have a constraint, but the symmetric difference (xor) of their labels is not empty" (the constraint cuts across boundaries).

Overall, I don't think labeling labels is much of a risk if we provide the tools to keep graph noodli-ness under control.

Reply:
Ok, I think this would be good safeguard.

By default, adding a constraint to an unlabeled node copies the other side's labels over. Then, have constraints between nodes with different labels (not incl. their primary ids) trigger warnings.

The user can configure some resource to disable them if they want.

I think there's a need to distinguish a "primary label" for systems, since their `SystemId` values aren't accessible using the standard builder API.

Reply:
> I think with that there's a need to distinguish a "primary label" for systems, since their SystemId values aren't accessible using the standard builder API.

We could expose this ;)